router/httpTriggers.go

```diff
@@ -162,6 +163,61 @@ func (ts *HTTPTriggerSet) watchTriggers() {
 	}
 }
 
+func (ts *HTTPTriggerSet) watchFunctions() {
+	rv := ""
+	for {
+		wi, err := ts.fissionClient.Functions(api.NamespaceAll).Watch(api.ListOptions{
+			ResourceVersion: rv,
+		})
+		if err != nil {
+			log.Fatalf("Failed to watch function list: %v", err)
+		}
+
+		for {
+			ev, more := <-wi.ResultChan()
+			if !more {
+				// restart watch from last rv
+				break
+			}
+			if ev.Type == watch.Error {
+				// restart watch from the start
+				rv = ""
+				time.Sleep(time.Second)
+				break
+			}
+			fn := ev.Object.(*tpr.Function)
+			rv = fn.Metadata.ResourceVersion
+
+			// update service map cache
+			for key := range ts.functionServiceMap.copy() {
+				if key.Name == fn.Metadata.Name &&
+					key.ResourceVersion != fn.Metadata.ResourceVersion {
+					err := ts.functionServiceMap.delete(key)
+					if err != nil {
+						log.Printf("Error deleting functionServiceMap cache: %v", err)
+					}
+					break
+				}
+
+			}
+
+			// update resolver function reference cache
+			for key, rr := range ts.resolver.copy() {
+				if key.functionReference.Name == fn.Metadata.Name &&
+					rr.functionMetadata.ResourceVersion != fn.Metadata.ResourceVersion {
+					err := ts.resolver.delete(key.namespace, &key.functionReference)
+					if err != nil {
+						log.Printf("Error deleting functionReferenceResolver cache: %v", err)
+					}
+					break
+				}
+			}
+
+			ts.syncTriggers()
```

Comment:
Need call ts.syncTriggers() to update functionHandler or the functionHandler will pass out-of-date function metadata to poolmgr. (https://github.com/fission/fission/blob/master/router/functionHandler.go#L42)

Reply:
Ah, yup, makes sense.